### PR TITLE
Rakesh.netalias

### PR DESCRIPTION
**Pull request by rakeshlacework**
Bug fix for network alias monitoring found in DO env. Pcap find devices reports all IP addresses configured for the physical interface. Network alias addresses are being reported along with the other since the alias points to the same physical interface. This breaks the assumption in code of one to one correspondence.

### CODE REVIEW

**Comment by miekg** [comment]
Thanks! Let me review this for a bit...

File: decode.go [hunk 1]
```diff
@@ -124,7 +125,7 @@ func (arp *Arphdr) String() (s string) {
 			decodemac(arp.SourceHwAddress), arp.SourceProtAddress,
 			decodemac(arp.DestHwAddress), arp.DestProtAddress)
 	} else {
-		s = fmt.Sprintf("addrtype = %d protocol = %d", arp.Addrtype, arp.Protocol)
+		s = "addrtype = " + strconv.FormatInt(int64(arp.Addrtype), 10) + " protocol = " + strconv.FormatInt(int64(arp.Protocol), 10)
```
Comment (miekg) [inline decode.go]: I don't understand this change, what's wrong with %d in fmt?

File: decode_test.go [hunk 2]
```diff
@@ -53,7 +53,7 @@ func TestDecodeSimpleTcpPacket(t *testing.T) {
 	if p.SrcMac != 0xbc305be8d349 {
 		t.Error("Src mac", p.SrcMac)
 	}
-	if len(p.Headers) != 2 {
+	if p.Headers_cnt != 2 {
```
Comment (miekg) [inline decode_test.go]: why does `len()` do the wrong thing here?

File: packet.go [hunk 3]
```diff
@@ -1,8 +1,10 @@
 package pcap
 
 import (
+	"bytes"
 	"encoding/binary"
 	"fmt"
+	"github.com/lacework/agent/datacollector/dlog"
```
Comment (miekg) [inline packet.go]: This add a 3rd party dep, so rather leave this out.

File: packet.go [hunk 4]
```diff
@@ -100,21 +122,32 @@ func (p *Packet) decodeArp() {
...
+		arp.DestHwAddress = pkt[8+arp.HwAddressSize+arp.ProtAddressSize : 8+2*arp.HwAddressSize+arp.ProtAddressSize]
+		arp.DestProtAddress = pkt[8+2*arp.HwAddressSize+arp.ProtAddressSize : 8+2*arp.HwAddressSize+2*arp.ProtAddressSize]
+
+		p.setHeader(arp)
 
-	p.Headers = append(p.Headers, arp)
-	p.Payload = p.Payload[8+2*arp.HwAddressSize+2*arp.ProtAddressSize:]
+		p.Payload = p.Payload[8+2*arp.HwAddressSize+2*arp.ProtAddressSize:]
+	} else {
+		dlog.Infof("s: %d L:%d(H:%d P:%d) -> T:%x D:%x S:%x A:%x P:%x %x", p.Seq, len(p.Payload), arp.HwAddressSize, arp.ProtAddressSize, p.Type, p.DestMac, p.SrcMac, arp.Addrtype, arp.Protocol, bytes.Split(p.Data, []byte(",")))
```
Comment (miekg) [inline packet.go]: I don't think logging will add much here, can't return an error due to the function's signature... so silently ignore?

File: packet.go [hunk 5]
```diff
@@ -100,21 +122,32 @@ func (p *Packet) decodeArp() {
...
+	}
 }
 
-func (p *Packet) decodeIp() {
+func (p *Packet) decodeIp(recur int) {
 	if len(p.Payload) < 20 {
 		return
 	}
+	if recur > 3 {
```
Comment (miekg) [inline packet.go]: should add some docs about this to the function

File: packet.go [hunk 6]
```diff
@@ -202,16 +240,21 @@ func (p *Packet) decodeIcmp() *Icmphdr {
...
 	icmp.Seq = binary.BigEndian.Uint16(pkt[6:8])
 	p.Payload = pkt[8:]
-	p.Headers = append(p.Headers, icmp)
+	p.setHeader(icmp)
+
 	return icmp
 }
 
-func (p *Packet) decodeIp6() {
+func (p *Packet) decodeIp6(recur int) {
```
Comment (miekg) [inline packet.go]: where is the recur limit enforced?

File: pcap.go [hunk 7]
```diff
@@ -2,17 +2,83 @@
...
+#cgo LDFLAGS: -Wl,-Bstatic -lpcap -Wl,-Bdynamic,--wrap=memcpy
 #include <stdlib.h>
 #include <pcap.h>
+#include <string.h>
+#include <ifaddrs.h>
 
-// Workaround for not knowing how to cast to const u_char**
-int hack_pcap_next_ex(pcap_t * p, struct pcap_pkthdr **pkt_header,
-		      u_char ** pkt_data)
+// See this for glibc 2.14 hack below
```
Comment (miekg) [inline pcap.go]: was the previous hack not working? I was less code is seems more portable?

File: pcap.go [hunk 8]
```diff
@@ -178,30 +271,56 @@ func (p *Pcap) Close() {
...
 
-func (p *Pcap) NextEx() (pkt *Packet, result int32) {
-	var pkthdr_ptr *C.struct_pcap_pkthdr
-	var pkthdr C.struct_pcap_pkthdr
+func (p *Pcap) NextEx(pktin *Packet) (pkt *Packet, result int32) {
+	if pktin == nil {
+		pkt = new(Packet)
+	} else {
+		pkt = pktin
+		pkt.Headers_cnt = 0
```
Comment (miekg) [inline pcap.go]: `Headers_cnt`: use CamelCase for this. I'm also not sure (as hinted above) about why do we need this? What was wrong with putting this in a slice and taking `len()`?

File: pcap.go [hunk 9]
```diff
@@ -285,21 +424,53 @@ func FindAllDevs() (ifs []Interface, err error) {
...
+		iface.Addresses, ipv4, ipv6 = findAllAddresses(dev.addresses)
+		if (ipv4 > 1 || ipv6 > 1) && getrc == 0 {
+			names := make(map[string]uint)
+			for k, ipaddr := range iface.Addresses {
+				iface.Addresses[k].IfName, iface.Addresses[k].Flags = getNetworkInterfaceInfo(ifaddrs, ipaddr)
+				if names[iface.Addresses[k].IfName] == 0 {
+					names[iface.Addresses[k].IfName] = iface.Addresses[k].Flags
+				}
+			}
+			for name, _ := range names {
```
Comment (miekg) [inline pcap.go]: lgtm, is there any way to test this?

File: pcap.go [hunk 10]
```diff
@@ -365,6 +543,7 @@ func (p *Pcap) DumpOpen(ofile *string) (dumper *PcapDumper, err error) {
 	return
 }
 
+/*
```
Comment (miekg) [inline pcap.go]: why is this commented out?

File: pcap_test.go [hunk 11]
```diff
@@ -55,6 +55,7 @@ func TestPcapOpenLive(t *testing.T) {
 	testPcapHandle(t, pcapOpenLive)
 }
 
+/*
```
Comment (miekg) [inline pcap_test.go]: why is this commented out?

File: tools/pass/pass.go [hunk 12]
```diff
@@ -9,7 +9,7 @@ import (
 	"io"
 	"os"
 
-	"github.com/miekg/pcap"
+	"github.com/lacework/pcap"
```
Comment (miekg) [inline tools/pass/pass.go]: ?

File: tools/tcpdump/tcpdump.go [hunk 13]
```diff
@@ -148,7 +149,7 @@ func Dumpline(addr uint32, line []byte) {
 	fmt.Print("  ")
 	for i = 0; i < 16 && i < uint16(len(line)); i++ {
 		if line[i] >= 32 && line[i] <= 126 {
-			fmt.Println("%c", line[i])
+			fmt.Printf("%c", line[i])
```
Comment (miekg) [inline tools/tcpdump/tcpdump.go]: seems a weird change?

**Comment by rakeshlacework** [comment]
Sorry not intended to send this, can you reject it

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/pcap] Rakesh.netalias (#..." ]
>Sorry not intended to send this, can you reject it

close the pr then

**Pull request closed without merging**